src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,232 @@
+    /// Both APIs require support for at least the following:
+    /// - `(R32_G32, Float)`
+    /// - `(R32_G32_B32, Float)`
+    /// - `(R16_G16, Float)`
+    /// - `(R16_G16_B16_A16, Float)`
+    /// - `(R16_G16, Inorm)`
+    /// - `(R16_G16_B16_A16, Inorm)`
+    /// VK could support more by querying `VK_FORMAT_FEATURE_ACCELERATION_STRUCTURE_VERTEX_BUFFER_BIT_KHR`, DX12 is not queryable? Note [the DX12 ray tracing spec](https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#d3d12_raytracing_geometry_triangles_desc) says it supports more than [the Win32 docs](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_geometry_triangles_desc).
+    pub vertex_format: Format,
+    pub vertex_buffer: &'a B::Buffer,
+    pub vertex_buffer_offset: Offset,
+    pub vertex_buffer_stride: Offset,
+
+    /// aka "vertex count"?
+    pub max_vertex: Offset,
+
+    // index format must be DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT
+    // can also be VK_INDEX_TYPE_NONE_KHR/DXGI_FORMAT_UNKNOWN
+    pub index_buffer: Option<(&'a B::Buffer, Offset, IndexType)>,
+
+    /// 3x4 matrix
+    /// TODO enum for cpu repr?
+    /// Must point to an address containing `TransformMatrix`
+    pub transform: Option<(&'a B::Buffer, Offset)>,
+}
+
+// VkTransformMatrixKHR
+#[derive(Debug)]
+pub struct TransformMatrix {
+    /// 3x4 row-major affine transformation matrix. Use `mint::RowMatrix3x4` if available.
+    pub matrix: [[f32; 4]; 3],
+}
+
+#[derive(Debug)]
+pub struct GeometryAabbs<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `AabbPositions`s
+    /// TODO: document lifetime required for this buffer
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+    pub buffer_stride: Offset,
+}
+
+// TODO doc
+#[derive(Debug)]
+pub struct AabbPositions {
+    /// Use `mint::Vector3` if available.
+    pub min: [f32; 3],
+    pub max: [f32; 3],
+}
+
+/// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryInstancesDataKHR.html
+#[derive(Debug)]
+pub struct GeometryInstances<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `Instance`s
+    /// TODO: document lifetime required for this buffer
+    /// TODO this struct also allows passing an array of pointers, idk if that makes sense outside the host operations case
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkGeometryInstanceFlagBitsKHR.html
+    pub struct InstanceFlags: u32 {
+        const TRIANGLE_FACING_CULL_DISABLE_BIT = 0x1;
+        const TRIANGLE_FRONT_COUNTERCLOCKWISE_BIT = 0x2;
+        const FORCE_OPAQUE_BIT = 0x4;
+        const FORCE_NO_OPAQUE_BIT = 0x8;
+    }
+}
+
+// TODO AFAIK rust doesn't have custom sized fields, so we'll need some binary writer wrapper to actually support this at an API level.
```

Comment:
Maybe [`packed_struct`](https://docs.rs/packed_struct/0.4.0/packed_struct/) or [Pre-RFC: Generic integers (uint<N> and int<N>)](https://internals.rust-lang.org/t/pre-rfc-generic-integers-uint-n-and-int-n/7641) could be applicable here?

Reply:
gfx-hal is a low-level API. It's fine to cut some corners and just document what bits are expected within `u32`.
Extra dependencies are discouraged

Reply:
My concern here is that we probably want this struct to be bit-for-bit compatible with [`VkAccelerationStructureInstanceKHR`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureInstanceKHR.html) and [`D3D12_RAYTRACING_INSTANCE_DESC`](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_instance_desc) so users can create an array of `acceleration_structure::Instance` and upload it directly to a GPU buffer. Otherwise, we need to settle for describing this carefully and letting users BYO packed struct impl... thoughts?

The Metal backend will probably have to do something special to repack into its own special format, however

Reply:
I did mean preserving the bit-by-bit if this needs to go into the driver. Just instead of doing bitfields, we could have raw unsigned integers with documented layouts - that's what I meant by cutting the corners.